### PR TITLE
[15.0][MIG] account_chart_update

### PR DESCRIPTION
**Pull request by flachica**

### CODE REVIEW

**Comment by MiquelRForgeFlow** [comment]
@flachica Currently, account_asset_management is in 14.0.2.6.1 version, so I think you should update the commit history of this PR accordingly 😉

**Comment by flachica** [comment]
After these Odoo [changes](https://github.com/odoo/odoo/commit/5419fe445dbaab24bdda1e7a5d25e4fac3fb2f1c) the tests is failing

I'm change [this](https://github.com/flachica/account-financial-tools/blob/15.0-mig-account_chart_update/account_chart_update/tests/test_account_chart_update.py#L93:L112) point but is not enought and I don't know if I'm in the correct way

Basically I think it boils down to the fact that now we do depend on having certain parts of a generic localization package installed. But I would like to confirm that I am correct. But it seems so far-fetched to me that I suppose there will be some simpler solution

Any help is welcome

**Comment by MiquelRForgeFlow** [comment]
@flachica please, cherry pick https://github.com/OCA/account-financial-tools/commit/0ac8e6045aa507d8d501c7828357d8411d27cc52

**Comment by flachica** [comment]
> @flachica please, cherry pick [0ac8e60](https://github.com/OCA/account-financial-tools/commit/0ac8e6045aa507d8d501c7828357d8411d27cc52)

Done

**Review by MiquelRForgeFlow — APPROVED** [review]

**Comment by Shide** [comment]
I found a functional error:
Steps to reproduce:
1. Start a clean database
2. Set company country to Spain
3. Install 10n_es
4. Install account_chart_update
5. Run straight account chart update
6. Clic next
7. Error happens:
```
One or more errors detected!

Updated account '100000 - Capital social'.
Updated account '100000 - Capital social'.
Updated account '103000 - Socios por desembolsos no exigidos, capital social'.
Updated account '103000 - Socios por desembolsos no exigidos, capital social'.
ERROR: Exception writing account '103000 - Socios por desembolsos no exigidos, capital social'.
Traceback (most recent call last):
  File "/opt/odoo/auto/addons/account_chart_update/wizard/wizard_chart_update.py", line 1026, in _update_accounts
    _logger.info(
  File "/usr/local/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/opt/odoo/custom/src/odoo/odoo/sql_db.py", line 135, in savepoint
    self.flush()
  File "/opt/odoo/custom/src/odoo/odoo/sql_db.py", line 108, in flush
    self.transaction.flush()
  File "/opt/odoo/custom/src/odoo/odoo/api.py", line 833, in flush
    env_to_flush['base'].flush()
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 5644, in flush
    process(self.env[model_name], id_vals)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 5635, in process
    recs._write(vals)
  File "/opt/odoo/custom/src/odoo/odoo/models.py", line 3949, in _write
    cr.execute(query, params + [sub_ids])
  File "<decorator-gen-3>", line 2, in execute
  File "/opt/odoo/custom/src/odoo/odoo/sql_db.py", line 89, in check
    return f(self, *args, **kwargs)
  File "/opt/odoo/custom/src/odoo/odoo/sql_db.py", line 310, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "mail_followers_mail_followers_res_partner_res_model_id_uniq"
DETAIL:  Key (res_model, res_id, partner_id)=(account.account, 0, 2) already exists.
```

File: account_chart_update/wizard/wizard_chart_update.py [hunk 1]
```diff
@@ -0,0 +1,1320 @@
...
+        :param str template: A template record.
+        :param str name: The name of the template model.
+        :return set: Fields to ignore in diff.
+        """
+        specials_mapping = {
+            "account.tax.template": {"chart_template_id", "children_tax_ids"},
+            "account.account.template": {"chart_template_id", "root_id", "nocreate"},
+            "account.fiscal.position.template": {"chart_template_id"},
+        }
+        specials = {
```
Comment (Shide) [inline account_chart_update/wizard/wizard_chart_update.py:667]:
```suggestion
        # Exclude `mail.thread` fields
        specials = set(self.env['mail.thread']._fields) | {
```

I've seen the problem is when trying to replace data on the `mail.message` fields.

I suggest to add all fields of the `mail.thread` model onto the fields to ignore. 
In previous versions, `account.account.template` doesn't inherit from `mail.thread`, and these fields makes not sense to override existing data. Maybe someone put some important messages on this model and we don't want to overwrite them.

**Review by Shide — CHANGES_REQUESTED** [review]

File: account_chart_update/wizard/wizard_chart_update.py (hunk 1 above)
Comment (flachica) [reply account_chart_update/wizard/wizard_chart_update.py:667]: Done

**Review by yajo — APPROVED** [review]
code review

**Comment by yajo** [comment]
The fix for the broken build is in https://github.com/OCA/account-financial-tools/pull/1360.

**Review by Shide — APPROVED** [review]

**Review by rafaelbn — APPROVED** [review]
Thanks 👍

**Comment by rafaelbn** [comment]
Please @flachica could you rebase? Because https://github.com/OCA/account-financial-tools/pull/1346#issuecomment-1084210780

**Comment by rafaelbn** [comment]
Please @flachica could you rebase?

**Comment by flachica** [comment]
> Please @flachica could you rebase?

Done

**Comment by rafaelbn** [comment]
Thank you @flachica , still failing test. Please check it if possible! Thank you very mucho for your contributions!

😄 

```
2022-04-04 11:28:40,516 257 ERROR odoo odoo.addons.account_chart_update.tests.test_account_chart_update: ERROR: TestAccountChartUpdate.test_chart_update
Traceback (most recent call last):
  File "/__w/account-financial-tools/account-financial-tools/account_chart_update/tests/test_account_chart_update.py", line 131, in setUp
    chart_by_company_user.try_loading()
  File "/opt/odoo/addons/account/models/chart_template.py", line 198, in try_loading
    template.with_context(default_company_id=company.id)._load(15.0, 15.0, company)
  File "/opt/odoo/addons/account/models/chart_template.py", line 283, in _load
    acc_template_ref, taxes_ref = self._install_template(company, code_digits=self.code_digits)
  File "/opt/odoo/addons/account/models/chart_template.py", line 568, in _install_template
    tmp1, tmp2 = self.with_context(lang=company.partner_id.lang)._load_template(company, code_digits=code_digits, account_ref=acc_ref, taxes_ref=taxes_ref)
  File "/opt/odoo/addons/account/models/chart_template.py", line 596, in _load_template
    generated_tax_res = self.with_context(active_test=False).tax_template_ids._generate_tax(company)
  File "/opt/odoo/addons/account/models/chart_template.py", line 1151, in _generate_tax
    raise UserError(_("Please first define a fiscal country for company %s.", company.name))
odoo.exceptions.UserError: Please first define a fiscal country for company Test account_chart_update company.
 
2022-04-04 11:28:40,516 257 INFO odoo odoo.addons.account_chart_update.tests.test_account_chart_update: Starting TestAccountChartUpdate.test_matching ... 
2022-04-04 11:28:40,989 257 INFO odoo odoo.addons.account_chart_update.tests.test_account_chart_update: ====================================================================== 
2022-04-04 11:28:40,989 257 ERROR odoo odoo.addons.account_chart_update.tests.test_account_chart_update: ERROR: TestAccountChartUpdate.test_matching
Traceback (most recent call last):
  File "/__w/account-financial-tools/account-financial-tools/account_chart_update/tests/test_account_chart_update.py", line 131, in setUp
    chart_by_company_user.try_loading()
  File "/opt/odoo/addons/account/models/chart_template.py", line 198, in try_loading
    template.with_context(default_company_id=company.id)._load(15.0, 15.0, company)
  File "/opt/odoo/addons/account/models/chart_template.py", line 283, in _load
    acc_template_ref, taxes_ref = self._install_template(company, code_digits=self.code_digits)
  File "/opt/odoo/addons/account/models/chart_template.py", line 568, in _install_template
    tmp1, tmp2 = self.with_context(lang=company.partner_id.lang)._load_template(company, code_digits=code_digits, account_ref=acc_ref, taxes_ref=taxes_ref)
  File "/opt/odoo/addons/account/models/chart_template.py", line 596, in _load_template
    generated_tax_res = self.with_context(active_test=False).tax_template_ids._generate_tax(company)
  File "/opt/odoo/addons/account/models/chart_template.py", line 1151, in _generate_tax
    raise UserError(_("Please first define a fiscal country for company %s.", company.name))
odoo.exceptions.UserError: Please first define a fiscal country for company Test account_chart_update company.
```

File: account_chart_update/wizard/wizard_chart_update.py [hunk 2]
```diff
@@ -0,0 +1,1321 @@
...
+    @api.model
+    @tools.ormcache("name")
+    def fields_to_ignore(self, name):
+        """Get fields that will not be used when checking differences.
+
+        :param str template: A template record.
+        :param str name: The name of the template model.
+        :return set: Fields to ignore in diff.
+        """
+        specials_mapping = specials = {
```
Comment (Shide) [inline account_chart_update/wizard/wizard_chart_update.py]: Why add `specials_mapping = specials`? You are overwritting it on L667

**Review by Shide — CHANGES_REQUESTED** [review]

File: account_chart_update/wizard/wizard_chart_update.py (hunk 2 above)
Comment (flachica) [reply account_chart_update/wizard/wizard_chart_update.py]: It is a typo. I push new code to solve it but tests still having problems. I need more time to try to solve it

**Review by Shide — APPROVED** [review]

**Comment by Shide** [comment]
I got an error:

Steps:
- Tested on Runboat
- Install l10n_es
- Duplicate 400000 and 600000 to 499999 and 699999 accounts and set it to Customer Invoices Journals and Vendor Bills Journals.
- Run wizard for PGCE PYMEs 2008 with Recreate missing XML-IDs

```
RPC_ERROR
Odoo Server Error


Traceback (most recent call last):
  File "/opt/odoo/odoo/tools/cache.py", line 85, in lookup
    r = d[key]
  File "/opt/odoo/odoo/tools/func.py", line 71, in wrapper
    return func(self, *args, **kwargs)
  File "/opt/odoo/odoo/tools/lru.py", line 34, in __getitem__
    a = self.d[obj]
KeyError: ('wizard.update.charts.accounts', <function WizardUpdateChartsAccounts.find_account_by_templates at 0x7f1c0900fc10>, account.account.template(account.account.template(275,),))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/odoo/api.py", line 886, in get
    return field_cache[record._ids[0]]
KeyError: account.account.template(275,)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/odoo/odoo/fields.py", line 1057, in __get__
    value = env.cache.get(record, self)
  File "/opt/odoo/odoo/api.py", line 889, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.account.template(account.account.template(275,),).code'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/opt/odoo/odoo/http.py", line 688, in dispatch
    result = self._call_function(**self.params)
  File "/opt/odoo/odoo/http.py", line 360, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/opt/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/opt/odoo/odoo/http.py", line 349, in checked_call
    result = self.endpoint(*a, **kw)
  File "/opt/odoo/odoo/http.py", line 917, in __call__
    return self.method(*args, **kw)
  File "/opt/odoo/odoo/http.py", line 536, in response_wrap
    response = f(*args, **kw)
  File "/opt/odoo/addons/web/controllers/main.py", line 1352, in call_button
    action = self._call_kw(model, method, args, kwargs)
  File "/opt/odoo/addons/web/controllers/main.py", line 1340, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 464, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/opt/odoo/odoo/api.py", line 451, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/mnt/data/odoo-addons-dir/account_chart_update/wizard/wizard_chart_update.py", line 374, in action_update_records
    self._update_taxes_pending_for_accounts(todo_dict)
  File "/mnt/data/odoo-addons-dir/account_chart_update/wizard/wizard_chart_update.py", line 1077, in _update_taxes_pending_for_accounts
    acc_id = self.find_account_by_templates(
  File "<decorator-gen-209>", line 2, in find_account_by_templates
  File "/opt/odoo/odoo/tools/cache.py", line 90, in lookup
    value = d[key] = self.method(*args, **kwargs)
  File "/mnt/data/odoo-addons-dir/account_chart_update/wizard/wizard_chart_update.py", line 534, in find_account_by_templates
    codes = templates.mapped("code")
  File "/opt/odoo/odoo/models.py", line 5447, in mapped
    recs = recs._fields[name].mapped(recs)
  File "/opt/odoo/odoo/fields.py", line 1176, in mapped
    self.__get__(first(remaining), type(remaining))
  File "/opt/odoo/odoo/fields.py", line 1083, in __get__
    recs._fetch_field(self)
  File "/opt/odoo/odoo/models.py", line 3276, in _fetch_field
    self._read(fnames)
  File "/opt/odoo/odoo/models.py", line 3343, in _read
    cr.execute(query_str, params + [sub_ids])
  File "<decorator-gen-3>", line 2, in execute
  File "/opt/odoo/odoo/sql_db.py", line 89, in check
    return f(self, *args, **kwargs)
  File "/opt/odoo/odoo/sql_db.py", line 310, in execute
    res = self._obj.execute(query, params)
  File "/opt/odoo-venv/lib/python3.8/site-packages/psycopg2/extensions.py", line 121, in getquoted
    pobjs = [adapt(o) for o in self._seq]
  File "/opt/odoo-venv/lib/python3.8/site-packages/psycopg2/extensions.py", line 121, in <listcomp>
    pobjs = [adapt(o) for o in self._seq]
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/odoo/odoo/http.py", line 644, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/opt/odoo/odoo/http.py", line 302, in _handle_exception
    raise exception.with_traceback(None) from new_cause
psycopg2.ProgrammingError: can't adapt type 'account.account.template'
```

**Comment by flachica** [comment]
> * set it to Customer Invoices and Vendor Bills.

Sorry for this newbie question, but how I set accounts to Customer Invoices and Vendor Bills?

**Comment by Shide** [comment]
> > * set it to Customer Invoices and Vendor Bills.
> 
> Sorry for this newbie question, but how I set accounts to Customer Invoices and Vendor Bills?

Sorry, i forgot to add "Journals" word at the end.

On the Journals, change the main accounts to other than 400000 and 600000

**Comment by flachica** [comment]
> * 699999

I could not replicate the error in local environment. But once I read the code, make some changes like [this](https://github.com/flachica/account-financial-tools/blob/cb3251041301ac1cf02358bb01d1b1f1a96fe8d2/account_chart_update/wizard/wizard_chart_update.py#L1078). I added .id to correct browse record.

**Review by rafaelbn — APPROVED** [review]

**Comment by rafaelbn** [comment]
- Customer Journals usually should have income accounts, in Spain something like 700000 or 705000
- Vendor Journals usually should have income accounts, in Spain something like 600000 or 629000

But in real life the accounts in customer or vendor journals are not used generally because the accounts come from the products (or product category) in the invoice lines.

**Review by Shide — APPROVED** [review]

**Comment by manuelregidor** [comment]
HI @flachica Do you think you'll be able to continue with this PR? Thank you.

**Comment by rafaelbn** [comment]
Hello @manuelregidor if we have 1 more reivew like you or @HaraldPanten we can merge this PR as is 😄 💪 Thank you!

**Review by HaraldPanten — APPROVED** [review]
It did work well for me in runbot.

LGTM

**Pull request merged**